src/type/result.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * Result型。新しいErrorクラスを作らなくてもエラーハンドリングできる。
+ *
+ * 成功時はsuccess関数を使ってSuccessResultを返し、
+ * エラー時はfailure関数を使ってFailureResultを返せば良い。
+ *
+ * 例：成功時はstringを返し、エラー時はErrorCodeA・ErrorCodeBを返す場合
+ *
+ * // 関数定義
+ * function generateString(): Result<string, "ErrorCodeA" | "ErrorCodeB"> {
+ *   // 成功時
+ *   return success("ok string");
+ *
+ *  // エラー時
+ *   return failure("ErrorCodeA", new Error("error message A"));
+ *   return failure("ErrorCodeB", new Error("error message B"));
+ *   return failure(new Error("other error"));
+ * }
+ *
+ * // 使う側
+ * const result = generateString();
+ * if (result.ok) {
+ *  // 成功時の処理
+ *   console.log(result.value); // -> "ok string"
+ * }
+ * else {
+ *   // エラー時の処理
+ *   console.log(result.code); // -> "ErrorCodeA" | "ErrorCodeB" | undefined
+ *   console.log(result.error); // -> Error
+ * }
+ *
+ * // getValueOrThrowを使えばエラー時にResultErrorをthrowし直す
+ * const value = getValueOrThrow(result);
+ * console.log(value); // -> "ok string"
+ *
+ * // Promise#thenと組み合わせると綺麗に書ける。
+ * const value = await asyncGenerateString().then(getValueOrThrow);
+ * console.log(value); // -> "ok string"
+ */
+export type Result<T, E extends string = string> =
+  | SuccessResult<T>
+  | FailureResult<E>;
```

Comment:
`code`が`string?`の場合エラーになってますね･･･ 😇 

とりあえずここを↓にして
```suggestion
export type Result<T, E extends string = string> =
  | SuccessResult<T>
  | FailureResult<E | undefined>;
```

`export class ResultError<E extends string = string> extends Error {`
のとこを
`export class ResultError<E extends string | undefined = string> extends Error {`
にすれば行ける感じでした。

が、どんなResultも`code=undefined`があり得るというコードになってしまう課題にも気づきました。
（以前はどんなResultも`"UNKNOWN"`があり得た）

型パズル難しいので、一旦こうで良い気がしています。。

Reply:
型エラーはなくせた…と思います